### ISSUE TITLE
There’s an exploitable re-entrancy issue in the stake function

### ISSUE THREAD

**Issue by SA124** [movementlabsxyz/movement#419]
Auditor: Ottersec
Code: MCR Settlement Contract

Issue that we believe could lead to loss of funds:
There’s an exploitable re-entrancy issue in the stake function. Since the passed in IERC20 custodian is unchecked, it can re-enter the same function during the [transferFrom](https://github.com/movementlabsxyz/movement/blob/f308ed27aa52867577a6ce3f10974ba8e52ff3de/protocol-units/settlement/mcr/contracts/src/staking/MovementStaking.sol#L289) call. Once stake calls transferFrom the contract will re-enter the stake with the valid/normal custody, the same stake amount and account. Only one real transfer of amount will be done but the [balance check](https://github.com/movementlabsxyz/movement/blob/f308ed27aa52867577a6ce3f10974ba8e52ff3de/protocol-units/settlement/mcr/contracts/src/staking/MovementStaking.sol#L292) will be passed in two separate stake function calls, duplicating the user’s staked balance. The user can then unstake and receive the tokens during rollover.

**Comment by l-monninger** [comment]
Yeah, so the intent here was to leave approving custodians up to MCR itself, i.e., manually reviewing. But, there also isn't a good case for this and several other methods to be reentrant. 

That being said, in the case linked, can't this only be exploited if the custodian token itself has a reentrancy bug in `transferFrom`?

**Comment by SA124** [comment]
<img width="411" alt="Screenshot 2024-08-22 at 2 50 10 PM" src="https://github.com/user-attachments/assets/4dbd0049-dea6-4bec-89ac-2be97b504f46">
<img width="351" alt="Screenshot 2024-08-22 at 2 50 26 PM" src="https://github.com/user-attachments/assets/3c35abf0-e528-428c-8870-10aa02ff152a">
<img width="354" alt="Screenshot 2024-08-22 at 2 50 40 PM" src="https://github.com/user-attachments/assets/b1fe5d0d-25ad-4ca3-a8cd-2f81f7c59b05">
<img width="353" alt="Screenshot 2024-08-22 at 2 50 52 PM" src="https://github.com/user-attachments/assets/3d4847d0-7e7b-425e-8e67-042cf11ee2c8">
<img width="352" alt="Screenshot 2024-08-22 at 2 51 04 PM" src="https://github.com/user-attachments/assets/253d1240-c724-49d5-bf81-7ddfc45b842a">
<img width="352" alt="Screenshot 2024-08-22 at 2 51 17 PM" src="https://github.com/user-attachments/assets/2210dd3d-a4ca-4ee2-b5ea-60f1082c0ac2">


@l-monninger

**Comment by SA124** [comment]
<img width="411" alt="Screenshot 2024-08-22 at 2 50 10 PM" src="https://github.com/user-attachments/assets/26910917-5d24-43b2-b843-dc07cf32995f">
<img width="351" alt="Screenshot 2024-08-22 at 2 50 26 PM" src="https://github.com/user-attachments/assets/a390d034-faaf-4faf-a6c7-2f871fb2d1f5">
<img width="354" alt="Screenshot 2024-08-22 at 2 50 40 PM" src="https://github.com/user-attachments/assets/92f574b6-d072-4ef0-b68b-305a2c6ee2a6">
<img width="353" alt="Screenshot 2024-08-22 at 2 50 52 PM" src="https://github.com/user-attachments/assets/cd40c243-2715-4fff-bf9b-a32b2dcf6e19">
<img width="352" alt="Screenshot 2024-08-22 at 2 51 04 PM" src="https://github.com/user-attachments/assets/427a8000-71fa-4227-963e-2c88353184d0">
<img width="352" alt="Screenshot 2024-08-22 at 2 51 17 PM" src="https://github.com/user-attachments/assets/c2b057bf-9a8c-444d-83a7-e4faa581f637">

@l-monninger

**Comment by l-monninger** [comment]
This should be covered by #501 . 

Also, marking as decentralized because MCR does not need to be safe for centralized launch.
